lib/code-caller-docker.js

```diff
@@ -0,0 +1,703 @@
+                // up before, but if it wasn't, we'll do so here to avoid a
+                // `mount point XXX is itself on a OSXFUSE volume` error.
+                try {
+                    await this.removeBindMount(this.hostDirectory.path);
+                } catch (e) {
+                    // If the above command failed, assume it because a mount point didn't
+                    // already exist and proceed as normal. If something is really messed
+                    // up, the next command will likely fail anyways.
+                }
+                await this.createBindMount(coursePath, this.hostDirectory.path);
+            } else {
+                throw new Error(`Unexpected worker execution mode "${config.workersExecutionMode}"`);
+            }
+
+        })(callback);
+    }
+
+    /**
+     * 
+     * @param {'restart' | 'question' | 'course-element' | 'core-element'} type 
+     * @param {string} directory 
+     * @param {string} file 
+     * @param {string} fcn 
+     * @param {string[]} args 
+     * @param {(err: Error | null | undefined, data?: any, output?: string) => void} callback 
+     */
+    call(type, directory, file, fcn, args, callback) {
+        this.debug('enter call()');
+        this.callCount += 1;
+        this.ensureChild().then(() => {
+            if (!this._checkState([WAITING])) return callback(new Error('invalid DockerCaller state'));
+            if (!this._checkReadyForCall()) return callback(new Error('not ready for call'));
+            // TODO: state checking might be off. for the 'course' allocation method, we only start
+            // containers when we know about the course. however, for the general case in prod, we can
+            // start containers immediately on boot by "warming up". Need to make sure we actually do that,
+            // and that it's handled correctly.
+
+            const callData = { type, directory, file, fcn, args };
+            const callDataString = JSON.stringify(callData);
+
+            this.output = '';
+            this.lastCallData = callData;
+            this.callback = callback;
+            // TODO: how long should timeout actually be?
+            // TODO: I think the best option here is to make this timeout longer (1.5x?)
+            // than the timeout in the worker. Starting python processes is cheap,
+            // but starting a new Docker container is relatively expensive. By allowing
+            // the container worker to timeout on bad python code, we'll receive that as
+            // a normal error during call(). The timeout at this level of the code should
+            // only kick in if a container goes completely unresponsive, at which point we
+            // have no choice but to kill it and start over.
+            // We can probably make the Docker worker take the timeout as an option so that
+            // we don't have to coordinate hardcoded timeouts in two places.
```

Comment:
It would be awesome to resolve this timeout TODO before deploying this to prod. We are currently fighting with timeout issues in prod with timeout issues from some course questions, so we are going to hit this issue as soon as we deploy 😄

Reply:
This should be resolved in e259532 - let me know if the implementation looks good to you!